### PR TITLE
do gc parallel

### PR DESCRIPTION
**Pull request by wentaoxu**
as title says

### CODE REVIEW

**Comment by wentaoxu** [comment]
PTAL @zhangjinpeng1987 @disksing @tangliu @ngaut

**Comment by zhangjinpeng1987** [comment]
Any test?

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -422,6 +423,127 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
...
+}
+
+type gcResult struct {
+	task           *gcTask
+	successRegions int
+	failedRegions  int
+	err            error
+}
+
+func getNextGCTask(store tikv.Storage, bo *tikv.Backoffer, safePoint uint64, lastKey kv.Key) (*gcTask, error) {
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: s/getNextGCTask/genNextGCTask

File: store/tikv/gcworker/gc_worker.go [hunk 2]
```diff
@@ -511,44 +633,30 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
...
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-	// Sleep to wait for all other tidb instances update their safepoint cache.
-	time.Sleep(gcSafePointCacheInterval)
-
-	log.Infof("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
-	startTime := time.Now()
+func doGCForOneRange(ctx goctx.Context, store tikv.Storage, task *gcTask, identifier string) *gcResult {
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Seems you should use a range as the input args from the function name.

File: store/tikv/gcworker/gc_worker.go [hunk 3]
```diff
@@ -422,6 +423,127 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
...
+
+			if task != nil {
+				go doGCTask(jobCtx, store, identifier, gcResultChan, task)
+				remained++
+			}
+		}
+	}
+}
+
+func doGCTask(ctx goctx.Context, store tikv.Storage, identifier string, resChan chan *gcResult, task *gcTask) {
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: This function should wait on a task queue and handle task one by one util receive nil.

File: store/tikv/gcworker/gc_worker.go (hunk 1 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok

File: store/tikv/gcworker/gc_worker.go [hunk 4]
```diff
@@ -422,6 +423,127 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
...
+	log.Infof("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
+	defer func() {
+		log.Infof("[gc worker] %s finish gc, running jobs %v, safePoint: %v, successful regions: %v,"+
+			" failed regions: %v, cost time: %s",
+			identifier, remained, safePoint, successRegions, failedRegions, time.Since(startTime))
+		gcHistogram.WithLabelValues("do_gc").Observe(time.Since(startTime).Seconds())
+	}()
+
+	var key []byte
+	gcResultChan := make(chan *gcResult, gcDefaultGCConcurrency)
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: gcTaskQueue := make(chan gcTask, gcDefaultGCConcurrency)
gcTaskCh := make(chan gcTask, 1024)
for i = 0; i < gcDefaultGCConcurrency; i++ {
 go runGcTaskWorker(gcTaskQueue, gcResultCh)
}
var key []byte
for {
  task, err := genNextGCTask(...)
  if err != nil {...}
  if task == nil { 
     send N nil to gcTaskQueue, notify all workers to quit
     break;
  }
  gcTaskQueue <- task

  ...receive task results if have
}
reveive task results util N nil

**Comment by ngaut** [comment]
Duplicate with https://github.com/pingcap/tidb/pull/5850

**Pull request closed without merging**